### ISSUE TITLE
Hybrid representation does not support or detect overlapping muffin tins

### ISSUE THREAD

**Issue by prckent** [QMCPACK/qmcpack#4381]
**Describe the bug**

The evaluation of the hybrid representation does not seem to support overlapping muffin tins. The code should either abort if any overlap or the code should be extended to support this case. Given the physical motivation behind the representation, the first seems easiest and should cover nearly all use cases.

**To Reproduce**

Non-variational energies can be found in VMC with large rcut

**Expected behavior**
Abort or consistent energies

**Comment by ye-luo** [comment]
Muffin tins are constructed with the effect of all the surroundings. I think evaluating orbitals based on the nearest center is solid.

**Comment by kayahans** [comment]
Here are the tests we ran on bilayer graphene and bilayer hBN. In the lmax convergence tests r_i is set to 1.4 for every species and in rcut tests lmax is set to 5. The graphs have error bars, but they are mostly barely visible. 

C-C bond length is 1.42 A, which would make r>1.34 Bohrs to get overlapping spheres. VMC energies in graphene become non-variational beyond rcut = 1.4 Bohrs. Similarly B-N bond length is 1.44 A that is where VMC energies become non-variational in BN as well. 

<img width="612" alt="image" src="https://user-images.githubusercontent.com/28604230/210637641-44877176-7d3a-400a-90a1-765345c3da97.png">
<img width="613" alt="image" src="https://user-images.githubusercontent.com/28604230/210637750-c1249361-dce4-4fd7-ae79-8367e52f7b82.png">
<img width="612" alt="image" src="https://user-images.githubusercontent.com/28604230/210637792-650d74b1-6473-4ba4-a975-1a8ba1fb6ed0.png">

**Comment by kayahans** [comment]
In bilayer graphene, I have looked at the portion of the data specifically to trace the origin of the error. Using mf-1.0 and lmax=5 data, I have plotted the kinetic energy and total energy as a function of rcut. Each energy are referenced to zero at rcut=0.6 Bohrs. 
![image](https://user-images.githubusercontent.com/28604230/210639060-f053b90d-36ba-4b77-be66-1392ee8e5ce2.png)

**Comment by ye-luo** [comment]
Non-variantional energy is a separate issue/feature. When the smooth scheme fail to make second derivative continuous, you do see energy kinks. Users needs to pick a proper imax and rcut to make the hybrid beneficial without hurting accuracy.

**Comment by jtkrogel** [comment]
What we are seeing is not due to the smoothing scheme.  The choice of nearest center in the case of overlapping spheres introduces step discontinuities in the wavefunction at the surfaces of the Voronoi polyhedra between the atomic centers.

In order to have variational energies, this discontinuity must be avoided.  The simplest way to avoid it is to abort with a message in the event of overlapping hybrid spheres.

**Comment by ye-luo** [comment]
For a given point in the overlapping area, the muffing tin of both sides are constructed based on the PW orbitals value. If you can accept the value discrepancy between one side muffin tin and PW, one should also accept the value discrepancy on the other side. Then the discrepancy between two muffin tins can be tolerated.

There is already discrepancy between muffin and intestinal area already. The data you provided cannot rule out the possibly that the discrepancy between muffin and intestinal area caused the issue. This can be rule out by stretching the lattice and see how the rcut behaves.

**Comment by ye-luo** [comment]
There is correlation between rcut and Lmax as well. A large rcut requires large lmax.

**Comment by jtkrogel** [comment]
We have two cases at least that demonstrate the onset of the error in the energy occurs at the point the spheres begin to overlap.  In the case of truly overlapping spheres, there is no smoothing at the boundary between them and true discontinuities are introduced.

**Comment by ye-luo** [comment]
> We have two cases at least that demonstrate the onset of the error in the energy occurs at the point the spheres begin to overlap. In the case of truly overlapping spheres, there is no smoothing at the boundary between them and true discontinuities are introduced.

Did you mean the figures above? There is a possibility that with rcut = 1.4 the discrepancy between discrepancy between muffin and intestinal area, already messed up the energy before even to worry about overlapping.

**Comment by jtkrogel** [comment]
@kayahans would you repost below the three energy vs rcut figures above (for C, B, and N) with a vertical line added at the point of sphere overlap?  This should clarify the issue.

**Comment by kayahans** [comment]
It looks like discontinuities start when the inner spheres touch each other, beyond the dashed lines. I had used the default of 0.3 bohrs for the interstitial region.

**Comment by ye-luo** [comment]
> It looks like discontinuities start when the inner spheres touch each other, beyond the dashed lines. I had used the default of 0.3 bohrs for the interstitial region.

What you show only explains energy jumps happen at rcut around half of the bound length but it is a not a direct proof of overlapping being the root cause.
1. Could you try graphite with lattice parameters doubled and see where energy jumps happen?
2. Use spo_scanner to scan the orbital values from one center to its nearest neighbor.
```
    Usage:
    add the following block at the end of determinantset block.
    <spo_scanner source="i">
      <path name="a" from_atom="0" to_atom="1" nknots="41"/>
      <path name="b" from_pos="0.0 0.0 0.0" to_pos="6.74632230 6.74632230 6.74632230" nknots="61"/>
    </spo_scanner>
```

**Comment by jtkrogel** [comment]
How about testing with fixed bond length and independently changing the B and N rmax's in hBN?  This way we could see if the energy jumps track the sphere overlap point (which would vary along the bond) or not.
